In [53]:
from pymatgen import Lattice, Structure
from pymatgen.symmetry.analyzer import SpacegroupAnalyzer
import pymatgen.symmetry.structure 
from jupyter_jsmol.pymatgen import quick_view
from pymatgen.ext.matproj import MPRester
from pymatgen.io.cif import CifWriter
import pymatgen.analysis.diffraction.xrd as xrd

In [54]:
with MPRester("pyNX4Po9ZxSuwfvq") as m:

    # Structure for material id
    structure = m.get_structure_by_material_id("mp-23210")

    w = CifWriter(structure)
    w.write_file('Structure-MgCl2.cif')
   

In [55]:
a = xrd.XRDCalculator(wavelength= 1.5406, symprec=0, debye_waller_factors=None)

In [56]:
pattern = a.get_pattern(structure, scaled=False, two_theta_range=(0, 90))
theta2 = pattern.x
intensity = pattern.y
hkls = pattern.hkls
dhkl = pattern.d_hkls



Les indices hkl, de même que les distances entre les plans d'une même famille hkl, sont déterminés à partir de la cellule primitive.

Les angles de diffraction sont déterminés à partir de la formule de bragg : $$2*d_{hkl}*sin(\theta) = n*\lambda$$

où dhkl est la distance entre les plans de la famille hkl, 𝜃 est l'angle de diffraction. 
Ici, n = 1 et 𝜆 =  1.5406 A.

Nous avons donc pour les 3 premiers pics : 



In [57]:
print("PREMIER PIC \n")
print("angle du premier pic  = ",pattern.x[0]/2,"°" )
print("indice hkl associé : ", pattern.hkls[0][0].get("hkl"))
print("distance entre les plans (1 1 1) : d_111 = ", dhkl[0],"A \n")


print("DEUXIEME PIC \n")
print("angle du deuxieme pic  = ",pattern.x[1]/2,"°" )
print("indice hkl associé : ", pattern.hkls[1][0].get("hkl"))
print("distance entre les plans (2 2 2) : d_222 = ", dhkl[1],"A \n")

print("TROISIEME PIC \n")
print("angle du troisieme pic  = ",pattern.x[2]/2,"°" )
print("indice hkl associé : ", pattern.hkls[2][0].get("hkl"))
print("distance entre les plans (1 0 0) : d_100 = ", dhkl[2],"A \n")


PREMIER PIC 

angle du premier pic  =  6.729022627826659 °
indice hkl associé :  (1, 1, 1)
distance entre les plans (1 1 1) : d_111 =  6.573995032059102 A 

DEUXIEME PIC 

angle du deuxieme pic  =  13.553169532697327 °
indice hkl associé :  (2, 2, 2)
distance entre les plans (2 2 2) : d_222 =  3.286997516029551 A 

TROISIEME PIC 

angle du troisieme pic  =  14.187876497310961 °
indice hkl associé :  (1, 0, 0)
distance entre les plans (1 0 0) : d_100 =  3.14277 A 



On peut vérifier que par exemple pour le premier pic : $2*d_{111}*sin(6.729022627826659) = 1.5406 A$